src/Open3D/Visualization/Rendering/Filament/FilamentRenderer.cpp

```diff
@@ -0,0 +1,272 @@
+
+    auto instanceId = resourceManager_.CreateMaterialInstance(id);
+
+    if (instanceId) {
+        auto wMaterialInstance =
+                resourceManager_.GetMaterialInstance(instanceId);
+        materialsModifier_->Init(wMaterialInstance.lock(), instanceId);
+    } else {
+        utility::LogWarning(
+                "Failed to create material instance for material handle {}.",
+                id);
+    }
+
+    return *materialsModifier_;
+}
+
+MaterialModifier& FilamentRenderer::ModifyMaterial(
+        const MaterialInstanceHandle& id) {
+    materialsModifier_->Reset();
+
+    auto wMaterialInstance = resourceManager_.GetMaterialInstance(id);
+    if (!wMaterialInstance.expired()) {
+        materialsModifier_->Init(wMaterialInstance.lock(), id);
+    } else {
+        utility::LogWarning(
+                "Failed to modify material instance: unknown instance handle "
+                "{}.",
+                id);
+    }
+
+    return *materialsModifier_;
+}
+
+void FilamentRenderer::RemoveMaterialInstance(
+        const MaterialInstanceHandle& id) {
+    resourceManager_.Destroy(id);
+}
+
+TextureHandle FilamentRenderer::AddTexture(const ResourceLoadRequest& request) {
+    if (request.path.empty()) {
+        request.errorCallback(request, -1,
+                              "Texture can be loaded only from file");
+        return {};
+    }
+
+    return resourceManager_.CreateTexture(request.path.data());
+}
+
+void FilamentRenderer::RemoveTexture(const TextureHandle& id) {
+    resourceManager_.Destroy(id);
+}
+
+IndirectLightHandle FilamentRenderer::AddIndirectLight(
+        const ResourceLoadRequest& request) {
+    if (request.path.empty()) {
+        request.errorCallback(request, -1,
+                              "Indirect lights can be loaded only from files");
+        return {};
+    }
+
+    return resourceManager_.CreateIndirectLight(request);
+}
+
+void FilamentRenderer::RemoveIndirectLight(const IndirectLightHandle& id) {
+    resourceManager_.Destroy(id);
+}
+
+SkyboxHandle FilamentRenderer::AddSkybox(const ResourceLoadRequest& request) {
+    if (request.path.empty()) {
+        request.errorCallback(request, -1,
+                              "Skyboxes can be loaded only from files");
+        return {};
+    }
+
+    return resourceManager_.CreateSkybox(request);
+}
+
+void FilamentRenderer::RemoveSkybox(const SkyboxHandle& id) {
+    resourceManager_.Destroy(id);
+}
+
+std::shared_ptr<RenderToBuffer> FilamentRenderer::CreateBufferRenderer() {
+    auto renderer = std::make_shared<FilamentRenderToBuffer>(engine_, *this);
+    bufferRenderers_.insert(renderer.get());
+    return std::move(renderer);
+}
+
+void FilamentRenderer::ConvertToGuiScene(const SceneHandle& id) {
+    auto found = scenes_.find(id);
+    if (found != scenes_.end()) {
+        // TODO: Warning on guiScene != nullptr
+
+        guiScene_ = std::move(found->second);
+        scenes_.erase(found);
+    }
+
+    // TODO: assert
```

Comment:
this TODO is under which condition? (if not found then LogError?)

Reply:
I have no idea, I've never looked at this function before. To me it looks like the code should be:
auto found = scenes_.find(id);
assert(found != scenes_.end());
if (found != scenes_.end()) { ... }
But since we can't assert, I guess that's why the TODO is there.
I've moved the TODO comment and added the warning in the other TODO.